core/src/main/java/io/grpc/internal/SerializingExecutor.java

```diff
@@ -135,39 +135,31 @@ public void execute(Runnable r) {
   private class TaskRunner implements Runnable {
     @Override
     public void run() {
-      boolean stillRunning = true;
-      try {
-        while (true) {
-          Runnable nextToRun;
+      Runnable nextToRun;
+      ArrayDeque<Runnable> tmp = null;
+      do {
+        if (tmp == null) {
           synchronized (internalLock) {
-            Preconditions.checkState(isThreadScheduled);
-            nextToRun = waitQueue.poll();
-            if (nextToRun == null) {
+            if (producerQueue.isEmpty()) {
               isThreadScheduled = false;
-              stillRunning = false;
-              break;
+              return;
             }
+            // Swap the queues rather than copying.
+            tmp = producerQueue;
+            producerQueue = consumerQueue;
           }
-
-          // Always run while not holding the lock, to avoid deadlocks.
+          consumerQueue = tmp;
+        }
+        while ((nextToRun = tmp.poll()) != null) {
```

Comment:
I'm suggesting to only use `tmp` when doing the swap. There is no need to use `tmp` here; just use `consumerQueue` directly. That is for readability because it means the reader doesn't have to follow the aliasing occurring with `tmp`.